xilem/src/view/virtual_scroll.rs

```diff
@@ -0,0 +1,355 @@
+                        view.rebuild(
+                            view,
+                            &mut state.state,
+                            ctx,
+                            widgets::VirtualScroll::child_mut(&mut element, *idx),
+                        );
+                    });
+                    state.requested_rebuild = false;
+                }
+            }
+        } else {
+            debug_assert!(view_state.cleanup_queue.is_empty());
+            for (&idx, child) in &view_state.previous_views {
+                if view_state.current_views.contains_key(&idx) {
+                    // We will handle this in the second loop
+                    continue;
+                }
+                ctx.with_id(view_id_for_index(idx), |ctx| {
+                    let state = view_state.view_states.get_mut(&idx).unwrap();
+                    child.teardown(
+                        &mut state.state,
+                        ctx,
+                        widgets::VirtualScroll::child_mut(&mut element, idx),
+                    );
+                    widgets::VirtualScroll::remove_child(&mut element, idx);
+                    view_state.cleanup_queue.push(idx);
+                });
+            }
+            for to_cleanup in view_state.cleanup_queue.drain(..) {
+                view_state.previous_views.remove(&to_cleanup).unwrap();
+                view_state.view_states.remove(&to_cleanup).unwrap();
+            }
+            for (idx, child) in view_state.current_views.drain() {
+                ctx.with_id(view_id_for_index(idx), |ctx| {
+                    if let Some(child_prev) = view_state.previous_views.get(&idx) {
+                        let state = view_state.view_states.get_mut(&idx).unwrap();
+                        child.rebuild(
+                            child_prev,
+                            &mut state.state,
+                            ctx,
+                            widgets::VirtualScroll::child_mut(&mut element, idx),
+                        );
+                        state.requested_rebuild = false;
+                        view_state.previous_views.insert(idx, child);
+                    } else {
+                        debug_assert!(used_action);
+                        let (new_child, child_state) = child.build(ctx);
+                        widgets::VirtualScroll::add_child(
+                            &mut element,
+                            idx,
+                            new_child.into_widget_pod(),
+                        );
+
+                        view_state.previous_views.insert(idx, child);
+                        view_state.view_states.insert(
+                            idx,
+                            private::ChildState {
+                                state: child_state,
+                                requested_rebuild: false,
+                            },
+                        );
+                    }
+                });
+            }
```

Comment:
I'd say the `if` statement doesn't clearly define what that condition is for, so I think a comment may be useful.